x-pack/plugin/shutdown/src/main/java/org/elasticsearch/xpack/shutdown/TransportDeleteShutdownNodeAction.java

```diff
@@ -16,26 +16,99 @@
+public class TransportDeleteShutdownNodeAction extends AcknowledgedTransportMasterNodeAction<Request> {
     private static final Logger logger = LogManager.getLogger(TransportDeleteShutdownNodeAction.class);
 
+    private static final DeleteShutdownNodeExecutor executor = new DeleteShutdownNodeExecutor();
+
+    private static boolean deleteShutdownNodeState(Map<String, SingleNodeShutdownMetadata> shutdownMetadata, Request request) {
+        if (shutdownMetadata.containsKey(request.getNodeId()) == false) {
+            // noop, the node has already been removed by the time we got to this update task
+            return false;
+        }
+
+        logger.info("removing shutdown record for node [{}]", request.getNodeId());
+        shutdownMetadata.remove(request.getNodeId());
+        return true;
+    }
+
+    private static void ackAndReroute(Request request, ActionListener<AcknowledgedResponse> listener, RerouteService rerouteService) {
+        rerouteService.reroute("node registered for removal from cluster", Priority.URGENT, new ActionListener<>() {
+            @Override
+            public void onResponse(ClusterState clusterState) {}
+
+            @Override
+            public void onFailure(Exception e) {
+                logger.warn(() -> "failed to reroute after deleting node [" + request.getNodeId() + "] shutdown", e);
+            }
+        });
+        listener.onResponse(AcknowledgedResponse.TRUE);
+    }
+
+    // package private for tests
+    record DeleteShutdownNodeTask(Request request, ActionListener<AcknowledgedResponse> listener, RerouteService rerouteService)
+        implements
+            ClusterStateTaskListener {
+        @Override
+        public void onFailure(Exception e) {
+            logger.error(new ParameterizedMessage("failed to delete shutdown for node [{}]", request.getNodeId()), e);
+            listener.onFailure(e);
+        }
+    }
+
+    // package private for tests
+    static class DeleteShutdownNodeExecutor implements ClusterStateTaskExecutor<DeleteShutdownNodeTask> {
+        @Override
+        public ClusterState execute(ClusterState currentState, List<TaskContext<DeleteShutdownNodeTask>> taskContexts) throws Exception {
+            var shutdownMetadata = new HashMap<>(getShutdownsOrEmpty(currentState).getAllNodeMetadataMap());
+            boolean changed = false;
+            for (final var taskContext : taskContexts) {
+                var request = taskContext.getTask().request();
+                try {
+                    changed |= deleteShutdownNodeState(shutdownMetadata, request);
+                } catch (Exception e) {
+                    taskContext.onFailure(e);
+                    continue;
+                }
+                var reroute = taskContext.getTask().rerouteService();
+                taskContext.success(taskContext.getTask().listener().delegateFailure((l, s) -> ackAndReroute(request, l, reroute)));
```

Comment:
Observation: this is definitely emerging boilerplate:

```
taskContext.success(taskContext.getTask().listener().delegateFailure((l, s) -> ...
```

At the moment not every task has an underlying `listener()` so we can't yet do this in a generic way, but many tasks do and the remainder should be fixable. I hope to be able to address that to the point where this noise goes away.

Reply:
Noted ;)